### DISCUSSION TITLE
"green lines" for beat divisors

### DISCUSSION THREAD

**Discussion by cihe13375** [ppy/osu#16767] [Ideas]
Currently the beat divisor is "global" in the editor, so after choosing a beat divisor, the snapping ticks applies to the whole map. This is causing some inconvenience when mapping songs with multiple beat divisors (e.g. 1/4 and 1/6 mixed):
(1) the mapper may accidently snap a note into an undesired beat divisor, and this is hard to detect sometimes (e.g. maybe I accidently snap a 1/4 into 1/3, but because 1/3 is also used a lot through the map so modding tools don't point it out);
(2) the length of slider may change unexpectedly if the sliderpath is edited/scaled under the "wrong" beat divisor, forcing the mapper to Ctrl+Z and go back to change beat divisor (or worse, leaving unsnapped sliders);
(3) in many cases (not all though) the same beat divisor can be used across multiple difficulties, so if we can "retain" beat divisors (but not object themselves) when creating new difficulties then we don't need to struggle with the beat divisor bar every time when working on new difficulties

These issues can be solved if we can save "local" beat divisor settings in a map. The idea is similar to the green lines (in stable), but they change beat divisors instead of SV or such. So a section of a song may look like this in the timeline:
![localbeatdivisors](https://user-images.githubusercontent.com/33959010/152428277-3dba2940-6b29-4f26-b141-793d58f29f1a.png)
So after initially creating these "green lines", the notes will be automatically snapped to the corresponding divisors in those sections. The local beat divisor settings are saved with the beatmap, and can be retained when creating new difficulties. This gives following benefits:
(1) find snap errors much easier (e.g. if there's anything incompatible with 1/4 in a 1/4 section it must be wrong);
(2) more convenience when mapping new difficulties or editing existing maps
(3) probably useful for features like beat-divisor based combo coloring

**Comment by bdach** [comment]
These would fit quite naturally in what is called "timing points" in lazer. Those already specify bpm and meter, so I don't see a huge reason not to specify snap divisor either.

That said I've [previously read](https://github.com/ppy/osu/discussions/16334#discussioncomment-1919267) that switching divisors during mapping frequently is customary and expected...?

**Reply by cihe13375** [reply]
Although the majority of songs only use one set of divisors (usually either 1/4 or 1/3), but I can say that there're also quite some songs (including many rhythm game music) switching a lot between different beat divisors.

**Comment by peppy** [comment]
Another option is to automatically grab the divisor from the last two objects. Is what we do with SV right now too, I believe.

**Reply by cihe13375** [reply]
It can work as long as the local divisors are saved and can be transplanted to new difficulties. Maybe need some extra consideration though (e.g. sometimes it can be preferred to use 1/4 all the time instead of actually switching to lower ones due to several 1/1 or 1/2 objects)

**Reply by peppy** [reply]
Sure. I'm still not sold on baking this into control points though. On the beat divisor control, we could simply have a button which reads something like "use current (1/4)" where `1/4` is the beat snap used for the last object.

I'm not sure about how useful having it transplanted across difficulties is, as you'd usually be using different snap divisors for different difficulties.

**Reply by cihe13375** [reply]
For transplantation across difficulties, I mainly refer to songs with 1/4 and 1/3 alternatively appears (because 1/3 will be 1/3 anyway). If it's just switching between the same divisor set (like 1/2/4) then transplantation is probably less useful.
